File: Novels/packages/6-mobile-android/StoryReader/app/src/main/java/com/storysphere/storyreader/model/ReadingChallenge.kt
package com.storysphere.storyreader.model

import java.time.OffsetDateTime

data class ReadingChallenge(
    val id: String,
    val name: String,
    val description: String?,
    val challengeType: String,
    val goal: Int,
    val goalType: String,
    val timeRange: String,
    val startDate: OffsetDateTime,
    val endDate: OffsetDateTime,
    val progress: Int,
    val status: String,
    val isPublic: Boolean
)

data class ChallengeParticipant(
    val userId: String,
    val progress: Int,
    val joinedAt: OffsetDateTime,
    val updatedAt: OffsetDateTime
)

data class ChallengeProgress(
    val challenge: ReadingChallenge,
    val participants: List<ChallengeParticipant>
)

data class ReadingGoal(
    val id: String,
    val goalType: String,
    val target: Int,
    val current: Int,
    val timeRange: String,
    val startDate: OffsetDateTime,
    val endDate: OffsetDateTime,
    val status: String
)

data class ActivityFeedItem(
    val id: String,
    val activityType: String,
    val timestamp: OffsetDateTime,
    val storyId: String?,
    val chapterId: String?,
    val metadata: Map<String, String>?
)

data class ActivityFeed(
    val items: List<ActivityFeedItem>,
    val total: Int,
    val page: Int,
    val limit: Int
)

data class ReadingStatistics(
    val activityCounts: List<ActivityCount>,
    val activeGoals: List<ReadingGoal>
) {
    data class ActivityCount(
        val activityType: String,
        val count: Int
    )
}

File: Novels/packages/5-mobile-ios/StoryReader/Model/ReadingChallenge.swift
import Foundation

struct ReadingChallenge: Identifiable, Decodable {
    let id: String
    let name: String
    let description: String?
    let challengeType: String
    let goal: Int
    let goalType: String
    let timeRange: String
    let startDate: Date
    let endDate: Date
    let progress: Int
    let status: String
    let isPublic: Bool
}

struct ChallengeParticipant: Identifiable, Decodable {
    var id: String { "\(userId)-\(joinedAt.timeIntervalSince1970)" }
    let userId: String
    let progress: Int
    let joinedAt: Date
    let updatedAt: Date
}

struct ChallengeProgressPayload: Decodable {
    let challenge: ReadingChallenge
    let participants: [ChallengeParticipant]
}

struct ActivityFeedPage: Decodable {
    let items: [ActivityFeedItem]
    let total: Int
    let page: Int
    let limit: Int
}

struct ActivityFeedItem: Identifiable, Decodable {
    let id: String
    let activityType: String
    let timestamp: Date
    let storyId: String?
    let chapterId: String?
    let metadata: [String: String]?
}

struct ReadingGoal: Identifiable, Decodable {
    let id: String
    let goalType: String
    let target: Int
    let current: Int
    let timeRange: String
    let startDate: Date
    let endDate: Date
    let status: String
}

struct ReadingStatistics: Decodable {
    struct ActivityCount: Decodable {
        let activityType: String
        let count: Int
    }
    
    let activityCounts: [ActivityCount]
    let activeGoals: [ReadingGoal]
}

